service/src/main/java/org/cbioportal/service/util/CoExpressionAsyncMethods.java

```diff
@@ -0,0 +1,68 @@
+package org.cbioportal.service.util;
+
+import org.springframework.stereotype.Component;
+import org.springframework.scheduling.annotation.Async;
+import org.cbioportal.model.CoExpression;
+import org.cbioportal.service.exception.GeneNotFoundException;
+import org.cbioportal.service.exception.GenesetNotFoundException;
+import org.apache.commons.lang3.math.NumberUtils;
+import org.apache.commons.math3.linear.Array2DRowRealMatrix;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.apache.commons.math3.stat.correlation.SpearmansCorrelation;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.math.BigDecimal;
+import java.util.concurrent.CompletableFuture;
+
+@Component
+public class CoExpressionAsyncMethods {
+
+    @Async
+    public CompletableFuture<CoExpression> computeCoExpression(String entityId, List<String> valuesA, List<String> valuesB,
+            Boolean isMolecularProfileBOfGenesetType, Double threshold, String molecularProfileId) throws GenesetNotFoundException, GeneNotFoundException {
+
+        List<String> valuesACopy = new ArrayList<>(valuesA);
+        List<String> valuesBCopy = new ArrayList<>(valuesB);
+
+        List<Integer> valuesToRemove = new ArrayList<>();
+        for (int i = 0; i < valuesBCopy.size(); i++) {
+            if (!NumberUtils.isNumber(valuesBCopy.get(i)) || !NumberUtils.isNumber(valuesACopy.get(i))) {
+                valuesToRemove.add(i);
+            }
+        }
+
+        for (int i = 0; i < valuesToRemove.size(); i++) {
+            int valueToRemove = valuesToRemove.get(i) - i;
+            valuesBCopy.remove(valueToRemove);
+            valuesACopy.remove(valueToRemove);
+        }
```

Comment:
I know you just copied this, but this whole section seems not great. I think this code is weird looking because we're trying to remove elements from the collection we're iterating over _and_ we're iterating over two collections simultaneously, so there's a nasty side effect. That said, we can still clean this up. Here's a helper function I came up with that accomplishes the same thing. You can use it if you'd like, just make sure you benchmark it if you do.

(Note the non short circuit or - I think this is the first time I've had to write one of those in like 2 years!)

```java
    private void removeNonNumbersFromBothLists(ArrayList<String> valuesACopy, ArrayList<String> valuesBCopy) {
        Iterator<String> itA = valuesACopy.listIterator();
        Iterator<String> itB = valuesBCopy.listIterator();
        while (itA.hasNext() && itB.hasNext()) {
            if (!NumberUtils.isCreatable(itA.next()) | !NumberUtils.isCreatable(itB.next())) {
                itA.remove();
                itB.remove();
            }
        }
    }
```